include/jni/unique_pointerlike.hpp

```diff
@@ -0,0 +1,83 @@
+#pragma once
+
+/*
+    A class like std::unique_ptr<T, D>, but holds T by value; operator->() returns T*,
+    not D::pointer; and deletion is performed via get_deleter()(current.Get()). These are
+    the semantics needed by:
+
+        * UniqueObject, a.k.a. UniquePointerlike<Object<TagType>, GlobalRefDeleter>,
+        * UniqueWeakObject, a.k.a. UniquePointerlike<Object<TagType>, WeakGlobalRefDeleter>,
+        * UniqueLocalObject, a.k.a. UniquePointerlike<Object<TagType>, LocalRefDeleter>,
+        * and similar typedefs for Array<E>.
+*/
+
+namespace jni
+   {
+    template < class T, class D >
+    class UniquePointerlike
+       {
+        private:
+            T pointerlike;
+            D deleter;
+
+            UniquePointerlike(const UniquePointerlike&) = delete;
+            UniquePointerlike& operator=(const UniquePointerlike&) = delete;
+
+        public:
+            UniquePointerlike()
+               : pointerlike(),
+                 deleter() {}
+
+            explicit UniquePointerlike(T&& t, D&& d)
+               : pointerlike(std::move(t)),
+                 deleter(std::move(d)) {}
+
+            UniquePointerlike(UniquePointerlike&& other)
+               : pointerlike(other.release()),
+                 deleter(std::move(other.get_deleter())) {}
+
+            template < class U >
+            UniquePointerlike(UniquePointerlike<U, D>&& other)
+               : pointerlike(other.release()),
+                 deleter(std::move(other.get_deleter())) {}
+
+            ~UniquePointerlike()
+               {
+                reset();
+               }
+
+            UniquePointerlike& operator=(UniquePointerlike&& other)
+               {
+                reset(other.release());
+                deleter = std::move(other.deleter);
+                return *this;
+               }
+
+            void reset(T&& t = T())
+               {
+                T current = pointerlike;
+                pointerlike = std::move(t);
+                if (current)
+                   {
+                    get_deleter()(current.Get());
+                   }
+               }
+
+            T release()
+               {
+                T current = pointerlike;
+                pointerlike = T();
+                return current;
+               }
+
+            explicit operator bool() const { return pointerlike; }
+
+            T* operator->() const          { return const_cast<T*>(&pointerlike); }
+            T& operator*() const           { return const_cast<T&>(pointerlike); }
```

Comment:
More options:

* Have `operator*` return `T` (by value copy). This wouldn't work for `operator->` though; it's the same issue as using `Object<>` as `unique_ptr`'s `pointer` type: you wouldn't be able to do `p->Get(...)`.
* Recapitulate the type punning we're using for the low level wrappers. Instead of _wrapping_ `jobject*` in `Object<>`, _cast_ `jobject*` to `Object<>*`. Then we can continue to use `unique_ptr` with custom deleters. Since `Object<>` et al have no data (other than the pointer) and no vtable, I _think_ this is safe, though unorthodox.

Reply:
* Delete `Object<>::operator=` ???

Reply:
> Compare `std::unique_ptr<T>::operator*`, which is `const` but returns non-const `T&`.

That's not a apt comparison: the `const`ness of that operator refers to the fact that the _container_ can't be changed (i.e. the pointer it holds cannot be replaced). A container like `std::unique_ptr<const T>` would still return a `const T&` with that operator.

Thinking about this more, I think it makes sense to implement the approach #32 suggests in combination with creating a family of holder types.

Reply:
My point was that `unique_ptr` does not propagate const-ness, and in general, pointer types don't propagate const-ness. You can mutate the pointed-to object through a const pointer; to my the pointed-to object itself const requires an additional `const` in a different place (`unique_ptr<const T>`, as you say).

I wound up fixing the hole by deleting `operator=`s and friending `UniquePointerlike` so that only it can update the pointer. I don't the reassignability was much used for `Object`/`Class`/`Array`, so this feels like a better solution than duplicating member functions into parallel class families, and then dealing with needing conversions between the two. It's also more backwards compatible, as it doesn't require a lot of rewriting `->` to `.` in client code.